Bluetooth: Controller: Fix Extended Advertising channel use

cvinayak opened the pull request:
Add implementation defined channel index in the auxiliary
pointer of the common extended payload format in the primary
channel PDUs and the same be used in the transmission of
auxiliary PDUs.

Fixes #35668.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix populate offset in latest advertising PDU

Fix implementation to populate the aux, and sync offset
in the latest PDU. If both the current and latest of the
double buffer has been filled and LLL did not pick the
latest PDU, then the offset should be filled into the latest
PDU (and not into the first/current PDU).

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix Extended Advertising channel use

Add implementation defined channel index in the auxiliary
pointer of the common extended payload format in the primary
channel PDUs and the same be used in the transmission of
auxiliary PDUs.

Fixes #35668.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add FIXME for Per Adv chain channel use

Add FIXME comments for missing use of channel selection
algorithm for Periodic Advertising chained PDUs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add advertiser clock accuracy value

Add implementation to set correct Advertiser's clock
accuracy value in the auxiliary pointer field in the common
extended payload format.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Minor rename of ULL internal function

Rename ULL internal helper function to get the random
address.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -65,6 +65,26 @@ static inline struct pdu_adv *lll_adv_scan_rsp_peek(struct lll_adv *lll)
 	return (void *)lll->scan_rsp.pdu[lll->scan_rsp.last];
 }
 
+static inline struct pdu_adv *lll_adv_pdu_latest_peek(struct lll_adv_pdu *pdu)
+{
+	uint8_t first;
+
+	first = pdu->first;
+	if (first != pdu->last) {
+		first += 1U;
+		if (first == DOUBLE_BUFFER_SIZE) {
+			first = 0U;
+		}
+	}
+
+	return (void *)pdu->pdu[first];
+}
```
It's a bit confusing that the function peeks the `latest` and then we use a variable named `first` to return the `latest`.

Thalley replied:
> first is the first valid (and currently used) PDU in the double buffer,

If that is the way you describe it, `first_valid` might have been a better choice of a name :) 

It isn't something that needs to be changed, but it is a interesting (and unconventional) naming

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -65,6 +65,26 @@ static inline struct pdu_adv *lll_adv_scan_rsp_peek(struct lll_adv *lll)
+
+	first = pdu->first;
+	if (first != pdu->last) {
+		first += 1U;
```
Just curious: Is there a reason why you are using `+= 1U` instead of `first++`?

Thalley replied:
> ;-) or `++first` (in C it does not matter, but following general conventions this is the version of incrementation that should be used in situations `x = x + 1`)

That's an old recommendation for old compilers though, right? 
IIRC a long time ago, `++i` and `i++` would result in different code, where `++i` would use one less register or something like that, but I think it's a long time since any reasonable compiler would do this.

ppryga replied:
Yes that is correct. I'm used to go for the pre-incrementation in situations like this.

cvinayak replied:
The only different I think of, is the size of `first` which is `uint8_t`, and by using `first += 1U;` the compiler has the choice to use a byte instead of a word to add LSB of the registers. Use of unary operator makes sense when the variable is machine word size, in which case an inc instruction will be used.

```
    uint8_t x = 5;
    
    x = x + 1;
```

versus

```
    int i;
    
    i++;
```

But I am not sure, what actually the compiler is doing.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -65,6 +65,26 @@ static inline struct pdu_adv *lll_adv_scan_rsp_peek(struct lll_adv *lll)
 	return (void *)lll->scan_rsp.pdu[lll->scan_rsp.last];
 }
 
+static inline struct pdu_adv *lll_adv_pdu_latest_peek(struct lll_adv_pdu *pdu)
```
```suggestion
static inline struct pdu_adv *lll_adv_pdu_latest_peek(const struct lll_adv_pdu *pdu)
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -65,6 +65,26 @@ static inline struct pdu_adv *lll_adv_scan_rsp_peek(struct lll_adv *lll)
+	return (void *)pdu->pdu[first];
+}
+
+static inline struct pdu_adv *lll_adv_data_latest_peek(struct lll_adv *lll)
```
```suggestion
static inline struct pdu_adv *lll_adv_data_latest_peek(const struct lll_adv *lll)
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -83,6 +103,12 @@ static inline struct pdu_adv *lll_adv_aux_data_peek(struct lll_adv_aux *lll)
 }
 
+static inline struct pdu_adv *
+lll_adv_aux_data_latest_peek(struct lll_adv_aux *lll)
```
```suggestion
lll_adv_aux_data_latest_peek(const struct lll_adv_aux *lll)
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -139,6 +165,12 @@ static inline struct pdu_adv *lll_adv_sync_data_peek(struct lll_adv_sync *lll,
 }
 
+static inline struct pdu_adv *
+lll_adv_sync_data_latest_peek(struct lll_adv_sync *lll)
```
```suggestion
lll_adv_sync_data_latest_peek(const struct lll_adv_sync *lll)
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -448,6 +441,53 @@ int ull_adv_aux_reset_finalize(void)
+					   uint8_t *const addr)
+{
+	if (addr) {
+		memcpy(addr, adv->rnd_addr, BDADDR_SIZE);
```
```suggestion
		(void)memcpy(addr, adv->rnd_addr, BDADDR_SIZE);
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1060,12 +1116,13 @@ void ull_adv_aux_offset_get(struct ll_adv_set *adv)
-struct pdu_adv_aux_ptr *ull_adv_aux_lll_offset_fill(uint32_t ticks_offset,
-						    uint32_t start_us,
-						    struct pdu_adv *pdu)
+struct pdu_adv_aux_ptr *ull_adv_aux_lll_offset_fill(struct lll_adv_aux *lll_aux,
```
`lll_aux` looks like it is unused in this function.

cvinayak replied:
Good catch.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1085,18 +1139,18 @@ struct pdu_adv_aux_ptr *ull_adv_aux_lll_offset_fill(uint32_t ticks_offset,
-		aux->offs = offs / (OFFS_UNIT_300_US / OFFS_UNIT_30_US);
-		aux->offs_units = 1U;
+		aux_ptr->offs = offs / (OFFS_UNIT_300_US / OFFS_UNIT_30_US);
+		aux_ptr->offs_units = 1U;
```
```suggestion
		aux_ptr->offs_units = SCA_VALUE_50_PPM;
```

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1085,18 +1139,18 @@ struct pdu_adv_aux_ptr *ull_adv_aux_lll_offset_fill(uint32_t ticks_offset,
-		aux->offs = offs;
-		aux->offs_units = 0U;
+		aux_ptr->offs = offs;
+		aux_ptr->offs_units = 0U;
```
```suggestion
		aux_ptr->offs_units = SCA_VALUE_500_PPM;
```

cvinayak committed:
Bluetooth: Controller: Explicitly typecast void return for memcpy calls

Explicitly typecast void return for memcpy calls.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Minor indentation fixes

Minor indentation fixes.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use defines for aux pointer offset unit value

Use defines for aux pointer offset unit value.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -2709,7 +2709,7 @@ static const uint8_t *adva_update(struct ll_adv_set *adv, struct pdu_adv *pdu)
 #if defined(CONFIG_BT_CTLR_ADV_EXT)
 		} else if (ll_adv_cmds_is_ext() && pdu->tx_addr) {
-			own_id_addr = ll_adv_aux_random_addr_get(adv, NULL);
+			own_id_addr = ull_adv_aux_random_addr_get(adv, NULL);
```
hmmm, good point. This function tried to follow `ll_addr_get`, but looks like the memcpy part of the branch is unused in the source base. I will check uses of `ll_addr_get` too and make specific functions instead of the now redundant branching inside the functions.

ppryga replied:
Are you going to do that in this PR?

cvinayak replied:
yes

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -65,6 +65,28 @@ static inline struct pdu_adv *lll_adv_scan_rsp_peek(struct lll_adv *lll)
 }
 
+static inline struct pdu_adv *
+lll_adv_pdu_latest_peek(struct lll_adv_pdu const *const pdu)
```
Isn't it more typical to have `const` as the first variable modifier, e.g. 
```suggestion
lll_adv_pdu_latest_peek(const struct lll_adv_pdu *const pdu)
```

This comment applies multiple places.

cvinayak replied:
Both coding styles are permitted.

Thalley replied:
They are, but I was just wondering if the suggestion from me is more in line with most other occurrences of `const` for the sake of being consistent in the code base.

cvinayak replied:
> This comment applies multiple places.

This change applies repo wide, hence should be taken up in separate PR and with wider audience.

Thalley replied:
> > This comment applies multiple places.
> 
> This change applies repo wide, hence should be taken up in separate PR and with wider audience.

Does that mean that the change you are introducing here in this PR should be moved to another PR, or are you suggesting that any revert of this would be done in a separate PR?

cvinayak replied:
I have moved out the `const` qualifier commit. Please re-review.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -81,8 +81,8 @@ uint8_t ll_adv_aux_random_addr_set(uint8_t handle, uint8_t const *const addr)
 	return 0;
 }
 
-uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref, uint8_t len,
```
This line isn't 100 characters though. 

80 is the recommendation, 100 is the max (which will be caught by CI afaik).

ppryga replied:
Yes it is just recommendation, so it is allowed to use longer line. 
What more Linux Kernel coding style does not enforce 100 characters either. It is a warning returned by kernel's checkpatch. Please see commit message in this PR: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=bdc48fa11e46f867ea4d75fa59ee87a7f48be144.

What more I'm not asking to do a change, I'm just point that this kind of updates is optional and up to committer preferences.

cvinayak committed:
Bluetooth: Controller: Separate address get and read functions

Have separate Bluetooth Device address get and read
functions, remove use of function just to return Extended
Advertising Random address and replace with simple
assignment statement.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

ppryga-nordic reviewed (commented):
One nic-picking comment. Besides looks LGTM.

Thalley commented on subsys/bluetooth/controller/ll_sw/ll_addr.c:
```diff
@@ -81,6 +60,31 @@ uint8_t ll_addr_set(uint8_t addr_type, uint8_t const *const bdaddr)
+	return pub_addr;
+}
+
+uint8_t *ll_addr_read(uint8_t addr_type, uint8_t *const bdaddr)
+{
+	uint8_t *addr;
+
+	addr = ll_addr_get(addr_type);
+	if (addr) {
+		memcpy(bdaddr, addr, BDADDR_SIZE);
+	}
+
+	return addr;
+}
```
Correct me if I am wrong, but this function seems to return the address, as well as store it in the `bdaddr` argument, effectively "returning" it twice. 

Is this by design? It's similar to the memcpy/memset functions, so it's not necessarily wrong, just making sure it is intentional.

cvinayak replied:
yes, it is by design, was simpler to return a valid pointer or NULL, compared to returning a errno.

Thalley commented on subsys/bluetooth/controller/ll_sw/ll_addr.c:
```diff
@@ -81,6 +60,31 @@ uint8_t ll_addr_set(uint8_t addr_type, uint8_t const *const bdaddr)
+
+	addr = ll_addr_get(addr_type);
+	if (addr) {
+		memcpy(bdaddr, addr, BDADDR_SIZE);
```
Should we have an check or an ASSERT ensuring that `bdaddr` is non-NULL?

cvinayak replied:
As the parameter was passed by internal module caller (hci.c or ull_scan.c) a NULL can be avoided.

Thalley reviewed (approved):
All my comments that needed to be addressed have been. 

A few final comments, but otherwise LGTM

ppryga-nordic reviewed (approved):
LGTM
